File: android/app/src/main/java/sol/vortexx/android/i18n/impl/AssetLocaleSource.kt
package sol.vortexx.android.i18n.impl

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import sol.vortexx.android.i18n.api.LocaleSource
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

private val Context.localePrefs by preferencesDataStore(name = "locale_prefs")
private val KEY_LOCALE = stringPreferencesKey("locale")

/**
 * 146-language loader. Locale JSON files live in `assets/locales/<code>.json`
 * — same shape as the web client's `/static/locales/*`. Lookup is:
 *   1. dotted path split ("nav.nodes" → ["nav", "nodes"])
 *   2. traverse nested JsonObject
 *   3. fall back to the en bundle if the key is missing
 *   4. if en is also missing, return the key itself so the UI is obvious
 *
 * Values are pre-parsed into a flat `Map<String, String>` on locale switch
 * so `translate()` stays O(1) — Compose recomposes many of them per frame.
 */
@Singleton
class AssetLocaleSource @Inject constructor(
    @ApplicationContext private val ctx: Context,
) : LocaleSource {

    private val json = Json { ignoreUnknownKeys = true }
    private val cache: MutableMap<String, Map<String, String>> = mutableMapOf()

    override val current: Flow<String> = ctx.localePrefs.data.map {
        it[KEY_LOCALE] ?: pickDeviceDefault()
    }

    override suspend fun setLocale(code: String) {
        ctx.localePrefs.edit { it[KEY_LOCALE] = code }
    }

    override suspend fun translate(key: String): String {
        val code = current.first()
        val bundle = loadBundle(code)
        bundle[key]?.let { return it }
        if (code != "en") loadBundle("en")[key]?.let { return it }
        return key
    }

    // ── internals ──────────────────────────────────────────────────────

    private fun loadBundle(code: String): Map<String, String> {
        cache[code]?.let { return it }
        val filename = "locales/$code.json"
        val flat = runCatching {
            ctx.assets.open(filename).bufferedReader().use {
                val obj = json.parseToJsonElement(it.readText()).jsonObject
                flatten(obj, prefix = "")
            }
        }.getOrDefault(emptyMap())
        cache[code] = flat
        return flat
    }

    private fun flatten(obj: JsonObject, prefix: String): Map<String, String> {
        val out = mutableMapOf<String, String>()
        for ((k, v) in obj) {
            val path = if (prefix.isEmpty()) k else "$prefix.$k"
            when (v) {
                is JsonObject -> out += flatten(v, path)
                else -> runCatching { out[path] = v.jsonPrimitive.content }
            }
        }
        return out
    }

    private fun pickDeviceDefault(): String {
        val lang = Locale.getDefault().language.lowercase()
        val country = Locale.getDefault().country.uppercase()
        return when {
            lang == "zh" && country == "TW" -> "zh-TW"
            lang.isNotBlank() -> lang
            else -> "en"
        }
    }
}

File: ios/Modules/Sources/I18N/impl/AssetLocaleSource.swift
import Foundation

/// 146-language loader. Locale JSON files live in `Resources/locales/`
/// (same shape as web client's `/static/locales/*.json` and Android's
/// `assets/locales/`).
///
/// Lookup per `translate(key)`:
///   1. `ns.key` split — traverse nested JSON.
///   2. Fall back to the `en` bundle if missing.
///   3. Return the key itself so missing strings are visually obvious.
public final class AssetLocaleSource: LocaleSource, @unchecked Sendable {
    private let bundle: Bundle
    private let defaults: UserDefaults
    private let lock = NSLock()
    private var cache: [String: [String: Any]] = [:]
    private let currentBridge = Bridge()

    public init(bundle: Bundle? = nil, defaults: UserDefaults = .standard) {
        self.bundle = bundle ?? .module
        self.defaults = defaults
        currentBridge.publish(defaults.string(forKey: "locale") ?? pickDeviceDefault())
    }

    public var current: AsyncStream<String> { currentBridge.stream() }

    public func setLocale(_ code: String) async {
        defaults.set(code, forKey: "locale")
        currentBridge.publish(code)
    }

    public func translate(_ key: String) async -> String {
        let code = currentBridge.lastValue()
        if let value = lookup(key, in: code) { return value }
        if code != "en", let fallback = lookup(key, in: "en") { return fallback }
        return key
    }

    // MARK: internals --------------------------------------------------

    private func lookup(_ key: String, in code: String) -> String? {
        let bundle = loadBundle(code)
        var cursor: Any = bundle
        for part in key.split(separator: ".") {
            guard let dict = cursor as? [String: Any],
                  let next = dict[String(part)] else { return nil }
            cursor = next
        }
        return cursor as? String
    }

    private func loadBundle(_ code: String) -> [String: Any] {
        lock.lock(); defer { lock.unlock() }
        if let existing = cache[code] { return existing }
        guard let url = bundle.url(forResource: code, withExtension: "json", subdirectory: "locales"),
              let data = try? Data(contentsOf: url),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { cache[code] = [:]; return [:] }
        cache[code] = obj
        return obj
    }

    private func pickDeviceDefault() -> String {
        let lang = (Locale.current.language.languageCode?.identifier ?? "en").lowercased()
        let region = (Locale.current.region?.identifier ?? "").uppercased()
        if lang == "zh" && region == "TW" { return "zh-TW" }
        return lang.isEmpty ? "en" : lang
    }

    private final class Bridge: @unchecked Sendable {
        private let lock = NSLock()
        private var subs: [UUID: AsyncStream<String>.Continuation] = [:]
        private var last: String = "en"
        func lastValue() -> String { lock.lock(); defer { lock.unlock() }; return last }
        func publish(_ v: String) {
            lock.lock(); last = v; let copy = subs.values; lock.unlock()
            for c in copy { c.yield(v) }
        }
        func stream() -> AsyncStream<String> {
            AsyncStream { cont in
                lock.lock(); cont.yield(last); let id = UUID(); subs[id] = cont; lock.unlock()
                cont.onTermination = { @Sendable _ in
                    self.lock.lock(); self.subs.removeValue(forKey: id); self.lock.unlock()
                }
            }
        }
    }
}
